### PR TITLE
Issue 13535: Handle Firefish chat messages

### LINKED ISSUE #13535: Chat messages from Firefisch are fragmented into
- [x] I have searched open and closed issues for duplicates

### Bug Description

I don't know if this is actually a bug. Therefore, feel free to categorise it differently if this is a mistake from me.

 - A chat with remote users can be started from Firefish. This is an @ addressed, private message that can be sent to a confirmed contact.
 - The receipt of the message is displayed on Friendica in the personal timeline. 
 - A reply can be created on Friendica and sent to Firefish.
 - The message and replies are displayed on Firefish in a thread (image 1)
 - If a new question is created and sent from Firefish, Friendica will not create a link between the first and the new message.
 - The new question is delivered in Friendica as a new, independent @ addressed, private message and displayed in the personal timeline (picture 2)
 - For the Friendica user, the connection between the messages is lost if further posts have been received in the personal timeline in the meantime.

### Steps to Reproduce

1. start a new chat from Firefish
2. select a Friendica user as chat partner
3. send a message as a Firefish user to the Friendica account
4. reply to this message as a Friendica user
5. as a Firefish user, reply again to the Friendica user's last reply.

Actual Result:

- Each new reply from Firefish to Friendica is treated as a new thread.
- The messages are sorted fragmented in the personal timeline

Expected Result:

- The messages are not fragmented
- The history of the conversation is traceable and is displayed in the correct order
- The message from Firefish is stored as a chat in the DM system?

### Screenshots
image 1
Screenshot 2023-10-14 at 06-24-18 tchncs
image 2
Screenshot 2023-10-14 at 06-26-41 ˈloma Network

### Platform Info

Friendica Version: dev

Friendica Source: git

PHP version: 8.1

SQL version: 10.5

### PR DESCRIPTION
**Pull request by annando**
Fixes #13535
Firefish chat messages lack the parent data. (Which I assume is intended)

To keep the chat messages together, we fetch the neccessary data by ourselves.

Also we now only store direct messages as such, when we are in a relation with this contact.

### CODE REVIEW

File: src/Protocol/ActivityPub/Processor.php [hunk 1]
```diff
@@ -1347,18 +1346,22 @@ public static function storeReceivers(int $uriid, array $receivers)
...
 	 */
-	private static function postMail(array $activity, array $item)
+	private static function postMail(array $item): bool
 	{
 		if (($item['gravity'] != Item::GRAVITY_PARENT) && !DBA::exists('mail', ['uri' => $item['thr-parent'], 'uid' => $item['uid']])) {
 			Logger::info('Parent not found, mail will be discarded.', ['uid' => $item['uid'], 'uri' => $item['thr-parent']]);
 			return false;
 		}
 
+		if (!Contact::isFollower($item['contact-id'], $item['uid']) && !Contact::isSharing($item['contact-id'], $item['uid'])) {
```
Comment (MrPetovan) [inline src/Protocol/ActivityPub/Processor.php:1360]: What about the `unkmail` setting where users can accept private messages even by people they don't follow or are followed by?

Comment (annando) [reply src/Protocol/ActivityPub/Processor.php:1360]: You cannot directly reply to messages you received that way. That setting is completely outdated.

Comment (MrPetovan) [reply src/Protocol/ActivityPub/Processor.php:1360]: Hmm.

**Review by MrPetovan — APPROVED** [review]

**Comment by nupplaphil** [comment]
@MrPetovan - you merged with failing tests :-/
--> I'll try to fix it with my PR

**Comment by MrPetovan** [comment]
The tests need the `php-intl` module to be enabled, not exactly sure how to go about that, but I know it was not in this PR.

**Pull request merged**